### PR TITLE
Return Access-Control-Expose-Headers in normal CORS requests

### LINKED ISSUE #33922: Access-Control-Expose-Headers not returned to request
### Describe the bug

The Access-Control-Expose-Headers is only returned during the preflight request but not included in the subsequent request itself. Consequently, the header cannot be accessed through JavaScript in the browser.

### Expected behavior

The Access-Control-Expose-Headers should be returned

### Actual behavior

The Access-Control-Expose-Headers  is not returned.

### How to Reproduce?

1 - Configure the quarkus.http.cors.exposed-headers with some header.
2 - Send a request from browser

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by rhuan080**
Fixes #33922

### CODE REVIEW

**Comment by sberyozkin** [comment]
@rhuan080 thanks for this PR, can you please add a test, update one of the existing Cors tests.

**Comment by rhuan080** [comment]
Hi @sberyozkin,

This header (and others) does not have a unit test. I'm not sure how to test it, as the handle method should be called.

**Comment by sberyozkin** [comment]
@rhuan080 I edited the PR title to make the bot happy.  Have a look at the Cors tests in the deployment module, CORSFullConfigHandlerTest, it has a few preflight request tests, copy one of them and verify this header is now set in the normal request case too

**Comment by rhuan080** [comment]
Thank you

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java [hunk 1]
```diff
@@ -176,6 +176,16 @@ public void handle(RoutingContext event) {
             if (allowedMethods != null) {
                 response.headers().add(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS, allowedMethods);
             }
+
+            //always set expose headers if present
+            if (exposedHeaders != null) {
```
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: @rhuan080 Is this block redundant, as you set this header below if it is not configured as a wildcard ?

**Comment by sberyozkin** [comment]
@rhuan080 Thanks for adding the test, I added a minor comment

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java (hunk 1 above)
Comment (rhuan080) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: @sberyozkin No, it is not. The exposedHeaders variable is the ACCESS_CONTROL_REQUEST_HEADERS coming from request, and the value is added. If the corsConfig.exposedHeaders has a value different from the wildcard, the corsConfig.exposedHeaders is used.

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: @rhuan080 

Why do the request headers have to be reported as exposed headers ?

FYI, the constructor has this line:
```
this.exposedHeaders = createHeaderString(this.corsConfig.exposedHeaders);
``` 

So it does look like to me it is a redundant block, am I still missing something ?

Comment (rhuan080) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: Hi, the wildcard allows all headers to be exposed, thus if the client sends the ACCESS_CONTROL_REQUEST_HEADERS it should be returned at Access-Control-Expose-Headers. But, if wildcard is not used and you use something like that (x-custom-header,y-custom-header) the Access-Control-Expose-Headers should not be returned, but the x-custom-header,y-custom-header configured.

Comment (rhuan080) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: Note that it is as the class was written. I'm just updating the Access-Control-Expose-Headers to be returned to no-prefight requests.

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: @rhuan080 Hi, I'm sorry if I'm a bit slow today, but I just don't follow this change.

>if the client sends the ACCESS_CONTROL_REQUEST_HEADERS it should be returned at Access-Control-Expose-Headers

Can you point me to the relevant spec text requiring it please ? As far as ACCESS_CONTROL_REQUEST_HEADERS is concerned, my understanding it is about the browser allowing certain headers be sent to the target by XHR/Fetch, while exposed headers are about telling to the browser which extra headers can be available in the response.

>Note that the class remains unchanged; I am solely updating the Access-Control-Expose-Headers to be returned for non-preflight requests.

Can you check my comment about your PR using essentially the same property twice ?

Comment (rhuan080) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: Hi @sberyozkin, I think you are right, but maybe it is another issue. Today, the Quarkus set the ACCESS_CONTROL_REQUEST_HEADERS to Access-Control-Expose-Headers when the wildcard (*) is used. However, it should set the response headers returned. Does that make sense? 

We can work on it on another issue. What do you think?

Comment (rhuan080) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: Hi, I agree! I will solve this

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: Thanks @rhuan080, it looks better,  but I thought it should be the bock you removed 
Which should stay, i.e, add the exposed headers only if they are non wildcard.

Can you please restore that one and remove the current one, and also, do the wikcard check in the constructor, similarly to how it is done for origins ? That would avoid doing this check on every request.

Nearly there, thanks 👍

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: I guess responding with a wildcard won't be a problem but it won't be of any help to the browser either ?

Comment (rhuan080) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:181]: Hi @sberyozkin,

The exposedHeaders variable has a calculated value from the createHeaderString method. This method already considers the wildcard ( return null to wildcard). Thus, I conclude that calling the isConfiguredWithWildcard method is unnecessary.

**Review by sberyozkin — APPROVED** [review]

**Comment by sberyozkin** [comment]
@rhuan080 Unfortunately we can't merge without the native security tests passing,  can you please rebase and push ?

**Comment by rhuan080** [comment]
Hi @roland-smith,
The spec says "An HTTP response to a [CORS request](https://fetch.spec.whatwg.org/#cors-request) that is not a [CORS-preflight request](https://fetch.spec.whatwg.org/#cors-preflight-request) can also include..". For me, it is not saying that the [CORS-preflight request](https://fetch.spec.whatwg.org/#cors-preflight-request) does not include this. header. Thus, it is not a bug.

**Comment by roland-smith** [comment]
Hi @rhuan080 as you said "[CORS-preflight request](https://fetch.spec.whatwg.org/#cors-preflight-request) does not include this header" and yet it is being included? how is that not a bug by your own words. 

I mean sure it might be ignored.. but its kinda messy leaving in.

**Comment by rhuan080** [comment]
Hi @roland-smith,

I said it is NOT saying that the [CORS-preflight request](https://fetch.spec.whatwg.org/#cors-preflight-request) does not include this header. It means it is not a bug.

**Comment by rhuan080** [comment]
Look at this spec words:

An HTTP response to a [CORS request](https://fetch.spec.whatwg.org/#cors-request) that is not a [CORS-preflight request](https://fetch.spec.whatwg.org/#cors-preflight-request) can **ALSO** include the following [header](https://fetch.spec.whatwg.org/#concept-header)

**Comment by roland-smith** [comment]
the words: _that **is not** a CORS-preflight request_  

would seem to say exactly that CORS-preflight does not include this header. 

if it was supposed to be included in the preflight response there would be utterly no point in adding an explicit section just for Access-Control-Expose-Headers. It would just be one of the headers that can be added to any CORS request. The spec would not need to explicitly call out a CORS preflight request.

**Comment by rhuan080** [comment]
@roland-smith 

Some headers should be used only for CORS preflight requests. The example is `[Access-Control-Allow-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers)` which is just used for CORS preflight requests. It is why the spec says [CORS request](https://fetch.spec.whatwg.org/#cors-request) that is not a [CORS-preflight request](https://fetch.spec.whatwg.org/#cors-preflight-request). It means it can be included in requests that are not an OPTIONS, but it is not saying the header should not be returned to CORS preflight requests (OPTIONS).

**Pull request merged**